Compilation error when using lifetime annotations with #[from]

Hi,

I’ve just run into an issue with thiserror. It may just be my inexperience with Rust’s lifetimes, though. The issue occurred when I was trying to use the `#[from]` attribute on an error that requires a lifetime specifier. I managed to create a minimal working example:

```rust
fn main() -> Result<(), Error<'static>> {
    Err(Error::SubError(SubError::Variant("some text")))
}

#[derive(Debug, thiserror::Error)]
enum Error<'a> {
    #[error("sub error")]
    SubError(#[from] SubError<'a>),
}

#[derive(Debug, thiserror::Error)]
enum SubError<'a> {
    #[error("variant: {0}")]
    Variant(&'a str),
}
```

This gives the following compilation error:

```
error[E0495]: cannot infer an appropriate lifetime due to conflicting requirements
 --> src/main.rs:5:17
  |
5 | #[derive(Debug, thiserror::Error)]
  |                 ^^^^^^^^^^^^^^^^
  |
note: first, the lifetime cannot outlive the lifetime `'a` as defined on the impl at 6:12...
 --> src/main.rs:6:12
  |
6 | enum Error<'a> {
  |            ^^
note: ...so that the types are compatible
 --> src/main.rs:5:17
  |
5 | #[derive(Debug, thiserror::Error)]
  |                 ^^^^^^^^^^^^^^^^
  = note: expected  `&Error<'_>`
             found  `&Error<'a>`
  = note: but, the lifetime must be valid for the static lifetime...
note: ...so that the type `SubError<'_>` will meet its required lifetime bounds
 --> src/main.rs:5:17
  |
5 | #[derive(Debug, thiserror::Error)]
  |                 ^^^^^^^^^^^^^^^^

error: aborting due to previous error
```

Any ideas as to how this could be solved?

Thanks for your work on anyhow and thiserror; IMHO they are the best the Rust ecosystem has to offer in terms of error handling :)

--- a/impl/src/valid.rs
+++ b/impl/src/valid.rs
@@ -2,7 +2,7 @@ use crate::ast::{Enum, Field, Input, Struct, Variant};
 use crate::attr::Attrs;
 use quote::ToTokens;
 use std::collections::BTreeSet as Set;
-use syn::{Error, Member, Result};
+use syn::{Error, GenericArgument, Member, PathArguments, Result, Type};
 
 impl Input<'_> {
     pub(crate) fn validate(&self) -> Result<()> {
@@ -181,6 +181,14 @@ fn check_field_attrs(fields: &[Field]) -> Result<()> {
             ));
         }
     }
+    if let Some(source_field) = source_field.or(from_field) {
+        if contains_non_static_lifetime(source_field) {
+            return Err(Error::new_spanned(
+                source_field.original,
+                "non-static lifetimes are not allowed in the source of an error",
+            ));
+        }
+    }
     Ok(())
 }
 
@@ -190,4 +198,23 @@ fn same_member(one: &Field, two: &Field) -> bool {
         (Member::Unnamed(one), Member::Unnamed(two)) => one.index == two.index,
         _ => unreachable!(),
     }
+}
+
+fn contains_non_static_lifetime(field: &Field) -> bool {
+    let ty = match field.ty {
+        Type::Path(ty) => ty,
+        _ => return false, // maybe implement later if there are common other cases
+    };
+    let bracketed = match &ty.path.segments.last().unwrap().arguments {
+        PathArguments::AngleBracketed(bracketed) => bracketed,
+        _ => return false,
+    };
+    for arg in &bracketed.args {
+        if let GenericArgument::Lifetime(lifetime) = arg {
+            if lifetime.ident != "static" {
+                return true;
+            }
+        }
+    }
+    false
 }

--- a/tests/ui/lifetime.rs
+++ b/tests/ui/lifetime.rs
@@ -1,0 +1,13 @@
+use thiserror::Error;
+
+#[derive(Error, Debug)]
+#[error("error")]
+struct Error<'a>(#[from] Inner<'a>);
+
+#[derive(Error, Debug)]
+#[error("{0}")]
+struct Inner<'a>(&'a str);
+
+fn main() -> Result<(), Error<'static>> {
+    Err(Error(Inner("some text")))
+}

--- a/tests/ui/lifetime.stderr
+++ b/tests/ui/lifetime.stderr
@@ -1,0 +1,5 @@
+error: non-static lifetimes are not allowed in the source of an error
+ --> $DIR/lifetime.rs:5:18
+  |
+5 | struct Error<'a>(#[from] Inner<'a>);
+  |                  ^^^^^^^^^^^^^^^^^

